modules/SofaBoundaryCondition/FixedRotationConstraint.inl

```diff
@@ -91,95 +91,52 @@ template <class DataTypes>
 void FixedRotationConstraint<DataTypes>::projectPosition(const core::MechanicalParams* /*mparams*/, DataVecCoord& xData)
 {
     helper::WriteAccessor<DataVecCoord> x = xData;
-    if (FixedXRotation.getValue() == true)
+    for (unsigned int i = 0; i < x.size(); ++i)
     {
-        for (unsigned int i = 0; i < x.size(); i++)
-        {
-            // Current orientations
-            sofa::defaulttype::Quat Q = x[i].getOrientation();
-
-            // Previous orientations
-            sofa::defaulttype::Quat Q_prev = previousOrientation[i];
-
-            Vec3 edgez, edgey_prev, edgex, edgey;
-            sofa::defaulttype::Mat<3, 3, Real > R;
-
-
-            edgex = Q.rotate(Vec3(1.0, 0.0, 0.0));
-            edgey_prev = Q_prev.rotate(Vec3(0.0, 1.0, 0.0));
-            edgez = cross(edgex, edgey_prev);
-            edgey = cross(edgez, edgex);
-            R[0][0] = edgex[0];    R[0][1] = edgex[1];    R[0][2] = edgex[2];
-            R[1][0] = edgey[0];    R[1][1] = edgey[1];    R[1][2] = edgey[2];
-            R[2][0] = edgez[0];    R[2][1] = edgez[1];    R[2][2] = edgez[2];
-
-            sofa::defaulttype::Quat newOrientation;
-            newOrientation.fromMatrix(R.transposed());
-            x[i].getOrientation() = newOrientation;
-
-            // Stores orientations for next iteration
-            previousOrientation[i] = newOrientation;
+        // Current orientations
+        sofa::defaulttype::Quat Q = x[i].getOrientation();
+        // Previous orientations
+        sofa::defaulttype::Quat Q_prev = previousOrientation[i];
+
+        auto project = [](const Vec3 a, const Vec3 b) -> Vec3 {
+            return (a * b) * b;
+        };
+        auto decompose_ts = [&](const sofa::defaulttype::Quat q, const Vec3 twistAxis) {
+            Vec3 vec3_part(q[0], q[1], q[2]);
+            Vec3 projected = project(vec3_part, twistAxis);
+            sofa::defaulttype::Quat twist(projected[0], projected[1], projected[2], q[3]);
+            // Singularity : A perpendicular angle would give you quaternion (0, 0, 0, 0)
+            if(std::none_of(twist.ptr(), twist.ptr() + 4 * sizeof(double), [](double x) {return x != 0. ;})) {
+                twist = sofa::defaulttype::Quat::identity();
+            }
+            twist.normalize();
+            sofa::defaulttype::Quat swing = q * twist.inverse();
+            swing.normalize();
+            return std::make_pair(twist, swing);
```

Comment:
Since you're using so much c++11/14/17 awesomeness in this PR I'd suggest replacing the std::make_pair(twist, swing) with std::pair(twist, swing) :)

Can't really judge the algo, although it seems like a good refactoring to me so +1

Reply:
Actually, I'm c++11 compliant ^^ What you are proposing will only work with c++17 deduction, right ?